### PR TITLE
Add the (basic) ability to persist data to Hyrise and use it in the TPC-H benchmark

### LINKED ISSUE #125: [ToDo] Remaining Implementation Tasks
An incomplete list of open tasks for the project implementation.
Feel free to edit.

- [ ] Decide how to align our approach with the old binary caching
  - What about the copied export() functions 
- [x] Decide if the Chunks should be able to persist themselves
  - Would be better for object orientation, but Chunks are pretty dump and should stay dump
- [ ] Write unit tests for implementation
- [ ] Improve object orientation in the Segment offset calculation
  - [ ] Move size calculation to the segments, `size_bytes()`
- [ ] Decide how to deal with the base vectors of the segments
  - Make them std::optional?

### PR DESCRIPTION
**Pull request by phoeinx**
## Main Contribution

This PR enables running the TPC-H Benchmark with mmap-based data.
To use memory-mapping, the `--use_mmap` flag has to be set. 

The rough program flow is as follows:
- Generate tables of `TPCHBenchmark`
- Exchange tables with mmap-based tables
   -  on `Chunk`-Level using `replace_chunk_with_persisted_chunk()` method of `StorageManager`
   - Persist the mmap-based tables to disk
- Run TPC-H Benchmark as usual

When the TPC-H tables already have been persisted:
- Map the saved tables from disk and create tables spanning over the mapped data
- Run TPC-H Benchmark as usual

### Most relevant changes

The most relevant changes happened in `StorageManager`. A good entry point for this PR is the `StorageManager::replace_chunk_with_persisted_chunk()` method. 

The persistence interface is now mainly offered by Table with 
- `Table::persist()` and
- `Table::persist_chunk(ChunkID chunk_id)`.

### Open questions
Most important questions:
For now I would like to focus the discussion on the architecture and interfaces:
- See [comment about OO](https://github.com/hyrise-mp-22-23/hyrise/pull/117#discussion_r1130758699).
-  Table-based persistence interface as described above. Does this make sense?

### Further contributions
This PR also implements:
- type-agnostic mapping of DictionarySegments (previously only DictionarySegments<int> were supported
- mapping of FixedStringDictionarySegments to and from memory region
- new mapping base using `byte *` pointer to avoid having to pad
- rudimentary persistence-file-table handling (being able to write chunks of a table to multiple files and handle references, as well as updating an already existing Chunk file)
- basic testing (which needs to be moved to real tests) -> the `--verify` option also proves correct implementation
- further refactoring

This Pull Request also includes the `storage.json`-PR #99  which enables the cold start of Hyrise.

### CODE REVIEW

File: src/lib/storage/storage_manager.cpp [hunk 1]
```diff
@@ -275,110 +439,93 @@ std::ostream& operator<<(std::ostream& stream, const StorageManager& storage_man
...
+  export_compressed_vector(*segment->compressed_vector_type(), *segment->attribute_vector(), file_name);
+  //TODO: What to do with non-compressed AttributeVectors?
 }
 
-void StorageManager::write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<int>> segment,
-                                                const std::string& file_name) {
+template <typename T>
+void StorageManager::_write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<T>> segment,
+                                                 const std::string& file_name) const {
   /*
```
Comment (phoeinx) [inline src/lib/storage/storage_manager.cpp:495]: remove as not up to date anymore

File: src/lib/storage/storage_manager.cpp [hunk 2]
```diff
@@ -275,110 +439,93 @@ std::ostream& operator<<(std::ostream& stream, const StorageManager& storage_man
...
   export_value(static_cast<uint32_t>(segment->attribute_vector()->size()), file_name);
 
-  export_values<int32_t>(*segment->dictionary(), file_name);
+  // we need to ensure that every part can be mapped with a uint32_t map
+  export_values<T>(*segment->dictionary(), file_name);
   export_compressed_vector(*segment->compressed_vector_type(), *segment->attribute_vector(), file_name);
+  //TODO: What to do with non-compressed AttributeVectors?
+}
+
+void StorageManager::_write_segment_to_disk(const std::shared_ptr<AbstractSegment> abstract_segment,
```
Comment (phoeinx) [inline src/lib/storage/storage_manager.cpp:511]: This is worth a discussion of the responsibilities of different classes/objects.

Should the StorageManager handle the whole persisting process and know detailed what to do per SegmentClass?

Or would it be nicer to let each Segment implement a `.byte_representation()` method, which would return a byte array to write out to disk. The Segments already know how to initialize (decode) themselves from a byte array and therefore should be able to encode themselves. 
+ this would keep knowledge about the internal structure in the Segments
+ this would decouple implementation and allow for easy extension

In conclusion, I would strongly argue for the second option.

File: src/lib/storage/storage_manager.cpp [hunk 3]
```diff
@@ -390,141 +537,204 @@ void StorageManager::write_chunk_to_disk(const std::shared_ptr<Chunk>& chunk,
...
-    chunk_segment_offset_ends[chunk_index] = segment_offset_ends;
-    chunk_offset_ends[chunk_index] = offset;
-  }
-  // Fill all offset fields, that are not used with 0s.
-  for (auto chunk_index = chunks.size(); chunk_index < StorageManager::_chunk_count; ++chunk_index) {
-    chunk_offset_ends[chunk_index] = uint32_t{0};
-  }
+std::pair<uint32_t, uint32_t> StorageManager::_persist_chunk_to_file(const std::shared_ptr<Chunk> chunk,
+                                                                     ChunkID chunk_id,
+                                                                     const std::string& file_name) const {
```
Comment (phoeinx) [inline src/lib/storage/storage_manager.cpp:548]: There is quite a bit of code duplication between persisting to an existing file and for a new file here. 
Either resolve code duplication or extract methods?

File: src/lib/storage/storage_manager.cpp [hunk 4]
```diff
@@ -390,141 +537,204 @@ void StorageManager::write_chunk_to_disk(const std::shared_ptr<Chunk>& chunk,
...
-  const auto file_bytes = std::filesystem::file_size(filename);
+  //Calls to mmap need to be pagesize-aligned
+  const auto pagesize = getpagesize();
+  const auto difference_to_pagesize_alignment = chunk_offset_begin % pagesize;
+  const auto page_size_aligned_offset = chunk_offset_begin - difference_to_pagesize_alignment;
 
-  // TODO: Remove unneccesary map on whole file
-  const auto* map = reinterpret_cast<uint32_t*>(mmap(NULL, file_bytes, PROT_READ, MAP_PRIVATE, fd, off_t{0}));
-  Assert((map != MAP_FAILED), "Mapping of File Failed.");
+  const auto* persisted_data =
```
Comment (phoeinx) [inline src/lib/storage/storage_manager.cpp:681]: consider renaming to `persisted_chunk` ?

File: src/lib/storage/storage_manager.cpp [hunk 5]
```diff
@@ -390,141 +537,204 @@ void StorageManager::write_chunk_to_disk(const std::shared_ptr<Chunk>& chunk,
...
 
     if (segment_index > 0) {
-      segment_offset_end = chunk_header.segment_offset_ends[segment_index - 1] + chunk_offset_end;
+      segment_offset_begin = chunk_header.segment_offset_ends[segment_index - 1];
     }
 
-    segments.emplace_back(std::make_shared<DictionarySegment<int32_t>>(map + segment_offset_end / 4));
+    resolve_data_type(column_definitions[segment_index], [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+      if constexpr (std::is_same<ColumnDataType, pmr_string>::value) {
```
Comment (phoeinx) [inline src/lib/storage/storage_manager.cpp:700]: Add comment that we can only persist `FixedStringDictionarySegments` and therefore know that this is one.

File: src/lib/storage/storage_manager.cpp (hunk 2 above)
Comment (Bouncner) [reply src/lib/storage/storage_manager.cpp:511]: 100% agree.

**Comment by Bouncner** [comment]
> Table-based persistence interface as described above. Does this make sense?

I guess the table itself does not write. Without looking at the code, what does `table->persist()` do?

File: src/lib/storage/storage_manager.cpp (hunk 2 above)
Comment (tjjordan) [reply src/lib/storage/storage_manager.cpp:505]: How do you like the way it is done now?

File: src/benchmarklib/abstract_table_generator.cpp [hunk 6]
```diff
@@ -342,11 +342,50 @@ void AbstractTableGenerator::generate_and_store() {
     std::cout << "- No indexes created as --indexes was not specified or set to false" << std::endl;
   }
 
+  _table_info_by_name = table_info_by_name;
+
+  // Persist tables
+  // As last step after completing encoding etc.
```
[the remark this answers is not in the document]
Comment (tjjordan) [reply src/benchmarklib/abstract_table_generator.cpp:348]: It this comment outdated?

File: src/lib/storage/storage_manager.cpp (hunk 2 above)
Comment (mweisgut) [reply src/lib/storage/storage_manager.cpp]: I agree with the second option and would vote against having the encoding/decoding logic for each segment type in the storage manager.

Alternatively, one could develop a BinarySegmentEncoded/-Decoder that encapsulates the encoding/decoding logic.

File: src/lib/storage/table.hpp [hunk 7]
```diff
@@ -225,7 +227,16 @@ class Table : private Noncopyable {
   const std::vector<ColumnID>& value_clustered_by() const;
   void set_value_clustered_by(const std::vector<ColumnID>& value_clustered_by);
 
+  /*
+   * Persist table to use mmap-based storage for its data.
+   * TODO: Explanation.
+   */
+
+  void persist();
```
Comment (mweisgut) [inline src/lib/storage/table.hpp:235]: Related to Martin's comment about not being responsible for persisting a table in the `Table` class: `persist_table` in the storage manager might be an alternative.

File: src/lib/storage/table.hpp [hunk 8]
```diff
@@ -225,7 +227,16 @@ class Table : private Noncopyable {
...
+
  protected:
+  void _persist_chunk(ChunkID chunk_id);
```
[the remark this answers is not in the document]
Comment (mweisgut) [reply src/lib/storage/table.hpp:238]: Do we have a use case that requires any component to persist/snapshot an individual chunk? Assuming we persist a table at time t1 and later at t2 again since the table has more chunks at t2, we want to have the functionality of persisting individual chunks that persist the delta instead of writing all chunks again. However, this could also be implemented as an optimization of `persist_table`. Then, we would call `persist_table` at t1 and t2, but at the second call, only the delta of immutable chunks will be persisted.

File: src/lib/storage/dictionary_segment.cpp [hunk 9]
```diff
@@ -137,8 +130,9 @@ ChunkOffset DictionarySegment<T>::size() const {
 template <typename T>
 std::shared_ptr<AbstractSegment> DictionarySegment<T>::copy_using_allocator(
     const PolymorphicAllocator<size_t>& alloc) const {
+  Assert(_dictionary_base_vector, "Cannot copy based on span-only DictionarySegment.");
```
[the remark this answers is not in the document]
Comment (tjjordan) [reply src/lib/storage/dictionary_segment.cpp:133]: We support it now :)
(Currently on the branch some_feedback)

File: src/lib/storage/table.cpp [hunk 10]
```diff
@@ -450,4 +455,16 @@ size_t Table::memory_usage(const MemoryUsageCalculationMode mode) const {
...
+void Table::persist() {
+  const auto chunk_count = _chunks.size();
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+    _persist_chunk(chunk_id);
+  }
+}
+
+void Table::_persist_chunk(ChunkID chunk_id) {
+  auto& storage_manager = Hyrise::get().storage_manager;
+  storage_manager.replace_chunk_with_persisted_chunk(_chunks[chunk_id], chunk_id, this);
```
Comment (tjjordan) [inline src/lib/storage/table.cpp:467]: Somewhere we need to check if the chunk is immutable.

File: src/lib/storage/table.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (tjjordan) [reply src/lib/storage/table.hpp:238]: For better object orientation, I think the chunks should be able to persist themselves.

File: src/lib/storage/storage_manager.cpp [hunk 11]
```diff
@@ -75,6 +230,15 @@ std::shared_ptr<Table> StorageManager::get_table(const std::string& name) const
   return table;
 }
 
+std::string StorageManager::_get_table_name(const Table* address) const {
+  for (const auto& [name, pointer] : _tables) {
+    if (pointer.get() == address) {
+      return name;
+    }
+  }
+  return std::string{};
```
[the remark this answers is not in the document]
Comment (tjjordan) [reply src/lib/storage/storage_manager.cpp:188]: Maybe no.

File: src/benchmark/tpch_benchmark.cpp [hunk 12]
```diff
@@ -50,13 +50,15 @@ int main(int argc, char* argv[]) {
     ("clustering", "Clustering of TPC-H data. The default of --clustering=None means the data is stored as generated "
                    "by the TPC-H data generator. With --clustering=\"Pruning\", the two largest tables 'lineitem' "
                    "and 'orders' are sorted by 'l_shipdate' and 'o_orderdate' for improved chunk pruning. Both are "
-                   "legal TPC-H input data.", cxxopts::value<std::string>()->default_value("None")); // NOLINT
+                   "legal TPC-H input data.", cxxopts::value<std::string>()->default_value("None")) // NOLINT
+    ("use_mmap", "Use storage.json to load TPCH data using the fileformat for chunks that hold their data with mmap."); // NOLINT
```
Comment (Bouncner) [inline src/benchmark/tpch_benchmark.cpp:54]: Site note: the user of the benchmarks/Hyrise should not care/be aware about the name of internal files that are used for persistence.

File: src/benchmarklib/abstract_table_generator.cpp [hunk 13]
```diff
@@ -342,11 +343,41 @@ void AbstractTableGenerator::generate_and_store() {
     std::cout << "- No indexes created as --indexes was not specified or set to false" << std::endl;
   }
 
+  /**
+   * Persist the tables in binary files if we want to use the new file format
```
Comment (Bouncner) [inline src/benchmarklib/abstract_table_generator.cpp:347]: Not of relevance: why not use the entire width of 120 chars for the comment?

File: src/benchmarklib/abstract_table_generator.cpp [hunk 14]
```diff
@@ -407,4 +438,52 @@ std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_loa
...
+    const auto persistence_data = mapping.second;
+    auto file_name = persistence_data.file_name;
+    const auto file_count = persistence_data.file_index;
+    auto total_chunks = std::vector<std::shared_ptr<Chunk>>{};
+
+    // Create column definitions
+    const auto column_definitions = storage_manager.get_table_column_definitions_from_json(table_name);
+
+    // Load chunks from files
+    for (auto index = size_t{0}; index < file_count + 1; ++index) {
```
Comment (Bouncner) [inline src/benchmarklib/abstract_table_generator.cpp:465]: Since the `+ 1` is not immediately clear (imo), I'd add a comment why you're using it here.

File: src/benchmarklib/abstract_table_generator.cpp [hunk 15]
```diff
@@ -407,4 +438,52 @@ std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_loa
...
+  for (const auto& mapping : tables_files_mapping) {
+    const auto table_name = mapping.first;
+
+    std::cout << "-  Loading table '" << table_name << "' from storage json. " << std::endl;
+    Timer timer;
+
+    const auto persistence_data = mapping.second;
+    auto file_name = persistence_data.file_name;
+    const auto file_count = persistence_data.file_index;
+    auto total_chunks = std::vector<std::shared_ptr<Chunk>>{};
```
Comment (Bouncner) [inline src/benchmarklib/abstract_table_generator.cpp:459]: strange name

Comment (Bouncner) [reply src/benchmarklib/abstract_table_generator.cpp:459]: Could be reserved().

File: src/benchmarklib/encoding_config.cpp [hunk 16]
```diff
@@ -4,7 +4,13 @@
 
 namespace hyrise {
 
-EncodingConfig::EncodingConfig() : EncodingConfig{SegmentEncodingSpec{EncodingType::Dictionary}} {}
+// Because it is hardly possible to save regular std:strings on disk, we use FixedStringDictionary as the default
```
Comment (Bouncner) [inline src/benchmarklib/encoding_config.cpp:7]: It's absolutely possible to persist them on disk, but doesn't help us here and needs some magic allocator stuff.

"hardly possible" is a strange wording. :)

(Not to be addressed)

File: src/benchmarklib/tpch/tpch_table_generator.cpp [hunk 17]
```diff
@@ -125,9 +126,24 @@ std::unordered_map<std::string, BenchmarkTableInfo> TPCHTableGenerator::generate
   Assert(_scale_factor < 1.0f || std::round(_scale_factor) == _scale_factor,
          "Due to tpch_dbgen limitations, only scale factors less than one can have a fractional part.");
 
-  const auto cache_directory = std::string{"tpch_cached_tables/sf-"} + std::to_string(_scale_factor);  // NOLINT
-  if (_benchmark_config->cache_binary_tables && std::filesystem::is_directory(cache_directory)) {
-    return _load_binary_tables_from_path(cache_directory);
+  auto cache_directory = std::string{};
+
+  if (_benchmark_config->use_mmap) {
+    if (!std::filesystem::is_directory("tpch_cached_tables_storage_json")) {
```
Comment (Bouncner) [inline src/benchmarklib/tpch/tpch_table_generator.cpp:132]: Not to be addressed: why use a flag `mmap` instead of simply passing a directory name that automatically states if we want to use persistency and which would avoid such hard-coded names?

File: src/lib/statistics/generate_pruning_statistics.cpp [hunk 18]
```diff
@@ -63,7 +63,10 @@ void generate_chunk_pruning_statistics(const std::shared_ptr<Chunk>& chunk) {
       if constexpr (std::is_same_v<SegmentType, DictionarySegment<ColumnDataType>>) {
         // we can use the fact that dictionary segments have an accessor for the dictionary
         const auto& dictionary = *typed_segment.dictionary();
-        create_pruning_statistics_for_segment(*segment_statistics, dictionary);
+        // DictionarySements use std::spans now, but this method still requires a std::vector. Because of this, we
+        // create a vector with the content of the span.
+        pmr_vector<ColumnDataType> dictionary_vector{dictionary.begin(), dictionary.end()};
+        create_pruning_statistics_for_segment(*segment_statistics, dictionary_vector);
```
Comment (Bouncner) [inline src/lib/statistics/generate_pruning_statistics.cpp:69]: What this part of the cold start measurements?

File: src/lib/storage/dictionary_segment.cpp [hunk 19]
```diff
@@ -30,76 +31,76 @@ template <typename T>
...
-      case PersistedSegmentEncodingType::DictionaryEncodingBitPacking: {
-        Fail("Support for span-based BitPackingVectors for DictionarySegments not implemented yet.");
-        break;
-      }
-      default: {
-        Fail("Unsupported EncodingType.");
-        break;
-      }
+  const auto header_data = reinterpret_cast<const uint32_t*>(start_address);
+  const auto encoding_type = PersistedSegmentEncodingType{header_data[ENCODING_TYPE_OFFSET_INDEX]};
```
Comment (Bouncner) [inline src/lib/storage/dictionary_segment.cpp:48]: Do we need this additional enum? All it stores has been available in  existing enums before?

File: src/lib/storage/dictionary_segment.cpp [hunk 20]
```diff
@@ -209,7 +207,26 @@ std::shared_ptr<const BaseCompressedVector> DictionarySegment<T>::attribute_vect
...
+  /*
+   * For a description of how dictionary segments look, see the following PR:
+   *    https://github.com/hyrise-mp-22-23/hyrise/pull/94
+   */
+  const auto compressed_vector_type_id =
+      StorageManager::resolve_persisted_segment_encoding_type_from_compression_type(compressed_vector_type().value());
+  StorageManager::export_value(static_cast<uint32_t>(compressed_vector_type_id), ofstream);
+
+  // Ee need to ensure that every part can be mapped with a uint32_t map.
+  StorageManager::export_value(static_cast<uint32_t>(dictionary()->size()), ofstream);
```
Comment (Bouncner) [inline src/lib/storage/dictionary_segment.cpp:224]: Not to be addressed: those methods should not be part of the stroage manager but rather some util class.

File: src/lib/storage/dictionary_segment.hpp [hunk 21]
```diff
@@ -94,15 +91,15 @@ class DictionarySegment : public BaseDictionarySegment {
   /**@}*/
 
  protected:
-  const std::shared_ptr<const pmr_vector<T>> _dictionary;
-  std::shared_ptr<const std::span<const T>> _dictionary_span;
+  const std::shared_ptr<const pmr_vector<T>> _dictionary_base_vector;
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/dictionary_segment.hpp:96]: I am fine with it.

File: src/lib/storage/fixed_string_dictionary_segment.cpp [hunk 22]
```diff
@@ -16,11 +17,75 @@ FixedStringDictionarySegment<T>::FixedStringDictionarySegment(
...
+      auto attribute_data_span = std::span<const uint8_t>(attribute_vector_address, attribute_vector_size);
+      auto attribute_vector = std::make_shared<FixedWidthIntegerVector<uint8_t>>(attribute_data_span);
+
+      _dictionary = dictionary_span_pointer;
+      _attribute_vector = attribute_vector;
+      _decompressor = _attribute_vector->create_base_decompressor();
+      break;
+    }
+    case PersistedSegmentEncodingType::DictionaryEncoding16Bit: {
+      auto* const attribute_vector_address =
```
Comment (Bouncner) [inline src/lib/storage/fixed_string_dictionary_segment.cpp:57]: Not to be addressed: There is a lot of duplication here and in the Dictionary class. That could  have been put into little helpers. There are also resolve patterns that would have helped to get rid of a lot of code here.

File: src/lib/storage/fixed_string_dictionary_segment.hpp [hunk 23]
```diff
@@ -70,13 +72,23 @@ class FixedStringDictionarySegment : public BaseDictionarySegment {
...
-  const std::shared_ptr<const BaseCompressedVector> _attribute_vector;
-  const std::unique_ptr<BaseVectorDecompressor> _decompressor;
+  const std::shared_ptr<const FixedStringVector> _dictionary_base_vector;
+  std::shared_ptr<const FixedStringSpan> _dictionary;
+  std::shared_ptr<const BaseCompressedVector> _attribute_vector;
+  std::unique_ptr<BaseVectorDecompressor> _decompressor;
+
+ private:
+  // Constants used for the construction of FixedStringDictionarySegments from memory-mapped storage.
+  static constexpr auto ENCODING_TYPE_OFFSET_INDEX = uint32_t{0};
```
Comment (Bouncner) [inline src/lib/storage/fixed_string_dictionary_segment.hpp:87]: Much much duplication duplication.

File: src/lib/storage/table.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/table.hpp:239]: I wouldn't argue that this is better object orientation. Apart from that: OO is soooooo 2000s! :D

**Review by Bouncner — COMMENTED** [review]
There is nothing that needs to be addressed this week. Just some remarks that might be of help in general or helps for paper discussions. Haven't looked at the storage manager yet.

**Review by tjeyy — APPROVED** [review]
